### PR TITLE
Delete redefinitions of methods in Variable already present on Tensor.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29907 Add more defensive tests for undefined tensors.
* #28287 Merge Tensor and Variable types.
* **#29667 Delete redefinitions of methods in Variable already present on Tensor.**
* #29665 Move most methods off Variable into torch::autograd::impl functions.
* #29653 Remove TensorImpl::is_variable, deprecate Tensor::is_variable

Some previous implementations are defined in native_functions.yaml.
In this case, I don't define them explicitly in Tensor; instead
they are placed in VariableTypeManual.cpp. When I did this, I would
have deleted documentation; instead, this documentation was moved
to native_functions.yaml

This also replaces `current_version` with just `_version`.

This is a carved out portion of #28287, rebased past Tensor-Variable
merge.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18504934](https://our.internmc.facebook.com/intern/diff/D18504934)

### CODE REVIEW

**Review by yf225 — APPROVED** [review]

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 21ad803:
* **1/1** broken upstream at merge base 6d54c5d (see [grid view](https://dr.pytorch.org/master-timeline.html?count=10&sha1=6d54c5ddd22ea69f6adaacc428d6c60aa63c8b77&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see [age history](https://dr.pytorch.org/master-viable-commits.html)):
      ```
      git fetch viable/strict
      git rebase viable/strict
      ```
* **0/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=21ad80374e3e93f67354dcce941e5902312b2b4f):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_xla_linux_xenial_py3_6_clang7_build | Build | `Failed to run '['/var/lib/jenkins/workspace/xla/scripts/generate_code.sh']'` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -37,19 +37,27 @@
   use_c10_dispatcher: full
   variants: function
 
+# Computes the gradient of current tensor w.r.t. graph leaves.
 - func: backward(Tensor self, Tensor? gradient=None, bool keep_graph=False, bool create_graph=False) -> ()
   variants: method
 
+# Sets the tensor data held by this `Variable` to be the same as `new_data`.
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: I think we should deprecate this in C++ (i.e., only in documentation).

You can say use something like NoGrad and in-place functions instead.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -2333,11 +2341,25 @@
 - func: sinh.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
   supports_named_tensor: True
 
+# Returns a copy of this `Variable` that is detached from its autograd graph
+# and has a blank version. This method is OK to call if the `Variable` is a
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: this isn't correct, I think you are confusing `detach` and `.data` (in python).

E.g.:
```
>>> torch.randn(2,3).add_(1).add_(2).detach()._version
2
>>> torch.randn(2,3).add_(1).add_(2).data._version
0
```

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: Well, in Python I can say `x.data = ...` and to the best of my knowledge we haven't deprecated that yet. Are you suggesting we deprecate this too?

Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml]: YES ! Please !
(note that might require update to some of our optimizers)

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: BTW, I plead "pre-existing" code ;) (I'll fix)

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -37,19 +37,31 @@
...
+# where Variables *are* Tensors (as opposed to them containing tensors, which
+# is what the previous interpretation was.)
 - func: set_data(Tensor(a!) self, Tensor new_data) -> ()
   use_c10_dispatcher: unboxed_only
   variants: method
 
 - func: data(Tensor self) -> Tensor
   variants: method
 
+# True if this `Variable` is a leaf and thus does not have a `grad_fn`.
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml]: Do we still call them `Variable`s in the comments? Or are we planning on calling everything Tensor in the future?

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -37,19 +37,31 @@
...
 - func: is_leaf(Tensor self) -> bool
   variants: method
 
+# Returns the input index of the gradient `Node` to which this
+# `Variable` is connected.  Note: input indexes of the gradient `Node`
+# correspond to output indexes of the corresponding forward `Node`.
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml]: "forward `None`" is not really a thing :/
Maybe just "forward function" here?

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -2333,11 +2345,25 @@
...
+# anymore, and in the `detach()` function we need to set `allow_tensor_metadata_change_`
+# to false to make such changes explicitly illegal, in order to prevent users from
+# changing metadata of the detached tensor and expecting the original tensor to also
+# be updated.
 - func: detach(Tensor self) -> Tensor
   use_c10_dispatcher: full
   supports_named_tensor: True
   variants: function, method
 
+# Like `detach()`, but removes this `Variable` in-place. This method may
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml]: "removes" -> "modifies" ?

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -37,19 +37,31 @@
   use_c10_dispatcher: full
   variants: function
 
+# Computes the gradient of current tensor w.r.t. graph leaves.
 - func: backward(Tensor self, Tensor? gradient=None, bool keep_graph=False, bool create_graph=False) -> ()
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml]: Does this method still exists? You remove it from Variable below?

File: torch/csrc/autograd/variable.cpp [hunk 7]
```diff
@@ -141,10 +84,6 @@ void Variable::remove_hook(unsigned pos) {
   (*list)[pos] = nullptr;
 }
 
-uint32_t Variable::current_version() const noexcept {
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp]: What is the reason for changing this "public looking" interface to the more "private looking" interface `_version()`? Should the user not be able to access the current version?

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: We should call everything Tensor, but that'll be a big codemod so I opted to minimize changes in this PR and have some places refer to things as Variable

File: torch/csrc/autograd/variable.cpp (hunk 7 above)
Comment (ezyang) [reply torch/csrc/autograd/variable.cpp]: Solely for parity with what the Python API looks like. And `current_version` was never supposed to be a public API, it was accidentally made so on Variable.

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: backward() always existed on Tensor; previously Variable overrode with its own definition (very naughtily, I might say: it's not a virtual function). When I delete on Variable, the old definition on Tensor is now the one exposed.

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: I plead preexisting comments XD (ok will fix)
